bite-size-notes/Week-3-d-Optimal Control-Jiangwen Cheng.md

```diff
@@ -1 +1,130 @@
+
+$$
+L_f b(x) + L_g b(x) u   \geq - 3
+$$
+
+
+![alt text](figs/1.png)
+
+2. As the system approaches the undesired set ${C}$, the range of available directions to take is reduced. 
+
+$$ 
+L_f b(x) + L_g b(x) u   > - 1 
+$$
+
+![alt text](figs/2.png)
+
+3. When the system is on the boundary of the undesired set, the only possible set of actions are to move along the boundary of ${C}$ or to move away in any direction.
+
+$$ 
+L_f b(x) + L_g b(x) u   \geq 0 
+$$
+
+![alt text](figs/3.png)
+
+4. When the system moves away from the undesired set, the possible range of actions in increased. 
+
+$$
+L_f b(x) + L_g b(x) u   \geq -1 
+$$
+
+![alt text](figs/4.png)
+
+## Optimization Based Control 
+We already established that Control Barrier Functions promise a necessary and sufficient conditions on safety. More important is that we wish to do so in a minimally invasive fashion, like modifying an existing controller in a minimal way so as to guarantee safety. This naturally leads to optimization based controllers:   
+Suppose we have a feedback controller $u = k(x)$ for the control system, and to modify this controller in a minimal way so as to guarentee safety, we can consider the following Quadratic Program (QP) based controller that finds the minimum perturbation on $u$: [3]
+
+$$
+\begin{align}
+u(x) = \text{argmin}_{u \in {R^m}} \hspace{0.3cm} \frac{1}{2} || u - k(x) ||_2^2 \\
+\\
+\hspace{0.5cm} \text{s.t.} \hspace{0.3cm} L_f b(x) + L_g b(x)
+u \geq \alpha (b(x)) 
+\end{align}
+$$  
+
+This approach combines control barrier functions (CBFs) with quadratic programming (QP) to compute control inputs that guarantee the invariance of a safe set defined by the CBFs. In practical applications, the QP-based controller acts as a safety filter over the nominal controller. It adjusts the control inputs to ensure they comply with the system's safety requirements. By doing so, it guarantees that the system remains within safe operational boundaries at all times. This setup is useful in dynamic environments that safety is extremely important, such as in autonomous vehicles and robotic systems. 
+
+
+## Conclusion 
+To summarize, this note touched on the foundational aspects of Control Lyapunov Functions (CLFs) and the concept of invariant sets, which are crucial for the discussion of Control Barrier Functions (CBFs). Then, we has discussed the fundamental mathematical principles of Control Barrier Functions (CBFs) and their practical application through CBF-based controllers in various real-life scenarios. These two are important in maintaining system stability and safety. Finally, we talked about CBF-QP controllers that can effectively act as safety filters, adjusting control inputs to ensure control system stay within safe boundaries. This capability is especially vital in dynamic fields. 
+
```

Comment:
Also, I am unable to view the entire document with the attached pictures after clicking on 'rich diff'. I checked with all the other notes and I was able to view notes under that section. For you I see it like this: 
![Screenshot 2024-04-24 at 9 46 41 PM](https://github.com/UW-CTRL/AA548-spr2024/assets/59079923/2a5ce5e8-246e-4cfa-9db3-bc5422495d10)

Before submitting the notes and merging it with the main, please do check it once if you are able to see the notes.

Reply:
My apology! I tried to solve the issue, but failed. I will try to figure it out later. And by the way, you could view the notes by clicking the three dot at the right corner. Thank you!